lib/ast/definition.cpp

```diff
@@ -118,7 +123,26 @@ SubsortMap kore_definition::get_subsorts() const {
     }
   }
 
-  return transitive_closure(subsorts);
+  subsorts_ = new SubsortMap(transitive_closure(subsorts));
+  return *subsorts_;
+}
+
+SubsortMap kore_definition::get_inverted_subsorts() {
+
+  if (subsorts_inverted_ != nullptr) {
+    return *subsorts_inverted_;
+  }
+
+  auto inverted_subsorts = SubsortMap{};
+
+  for (auto const &[inner, outers] : get_subsorts()) {
+    for (auto *outer : outers) {
+      inverted_subsorts[outer].insert(inner);
+    }
+  }
+
+  subsorts_inverted_ = new SubsortMap(inverted_subsorts);
```

Comment:
@theo25, I'm not sure if we need `transitive_closure` here. Could you please advise?

Reply:
The `get_subsorts()` method has entries:
```
S |-> {T . S is a subsort of T}
```
That is, it tells you _every_ sort that's greater than `S`, including transitively.

I think in your specific case you only need the non-transitive entries:
```
S |-> {T . S is a direct supersort of T}
```
such that `S` having a non-empty entry tells us that it has at least one subsort.

For consistency, I'd probably prefer to make your implementation transitive as well, but if you choose not to do this please make sure to document its contract similarly to `get_subsorts()`.

Reply:
I see, and it makes sense! The problem here is that the version I have cached contains the transitive case. I'm unsure if we gain a lot caching only this version as the `transitive_closure` function seems pretty expensive to me.

Reply:
Should we cache both versions or just keep the non-transitive version, @Baltoli?